src/calendarimportmodel.cpp

```diff
@@ -0,0 +1,240 @@
+                qWarning() << "Invalid notebook UID" << notebookUid;
+                return false;
+            }
+        }
+    }
+
+    if (importFromFile(mFileName, calendar))
+        storage->save();
+
+    storage->close();
+
+    return true;
+}
+
+QHash<int, QByteArray> NemoCalendarImportModel::roleNames() const
+{
+    QHash<int, QByteArray> roleNames;
+    roleNames[DisplayLabelRole] = "displayLabel";
+    roleNames[DescriptionRole] = "description";
+    roleNames[StartTimeRole] = "startTime";
+    roleNames[EndTimeRole] = "endTime";
+    roleNames[AllDayRole] = "allDay";
+    roleNames[LocationRole] = "location";
+    roleNames[UidRole] = "uid";
+
+    return roleNames;
+}
+
+bool NemoCalendarImportModel::importFromFile(const QString &fileName,
+                                             KCalCore::Calendar::Ptr calendar)
+{
+    QString filePath;
+    QUrl url(fileName);
+    if (url.isLocalFile())
+        filePath = url.toLocalFile();
+    else
+        filePath = fileName;
+
+    if (!(filePath.endsWith(".vcs") || filePath.endsWith(".ics"))) {
+        qWarning() << "Unsupported file format" << filePath;
+        return false;
+    }
+
+    QFile file(filePath);
+    if (!file.open(QIODevice::ReadOnly | QIODevice::Text)) {
+        qWarning() << "Unable to open file for reading" << filePath;
+        return false;
+    }
+    QString fileContent(file.readAll());
+
+    bool ok = false;
+    if (filePath.endsWith(".vcs")) {
+        KCalCore::VCalFormat vcalFormat;
+        ok = vcalFormat.fromString(calendar, fileContent);
+    } else if (filePath.endsWith(".ics")) {
+        KCalCore::ICalFormat icalFormat;
+        ok = icalFormat.fromString(calendar, fileContent);
+    }
+    if (!ok)
+        qWarning() << "Failed to import from file" << filePath;
+
+    return ok;
+}
+
+static bool incidenceLessThan(const KCalCore::Incidence::Ptr e1,
+                              const KCalCore::Incidence::Ptr e2)
+{
+    if (e1->dtStart() == e2->dtStart()) {
+        int cmp = QString::compare(e1->summary(),
+                                   e2->summary(),
+                                   Qt::CaseInsensitive);
+        if (cmp == 0)
+            return QString::compare(e1->uid(), e2->uid()) < 0;
+        else
+            return cmp < 0;
+    } else {
+        return e1->dtStart() < e2->dtStart();
+    }
+}
+
+bool NemoCalendarImportModel::importToMemory(const QString &fileName)
+{
+    if (!mEventList.isEmpty())
+        mEventList.clear();
+
+    beginResetModel();
+    KCalCore::MemoryCalendar::Ptr cal(new KCalCore::MemoryCalendar(KDateTime::Spec::LocalZone()));
+    importFromFile(fileName, cal);
+    KCalCore::Incidence::List incidenceList = cal->incidences();
+    for (int i = 0; i < incidenceList.size(); i++) {
+        KCalCore::Incidence::Ptr incidence = incidenceList.at(i);
+        if (incidence->type() == KCalCore::IncidenceBase::TypeEvent)
+            mEventList.append(incidence.staticCast<KCalCore::Event>());
```

Comment:
To be noted that exceptions, i.e. instances having recurrence id, on recurring events will show up as separate incidences on this model.